File: easy/9. Palindrome Number/solution.cs
/*
 * Solution for the 'Palindrome Number' problem.
 * Time complexity: O(log(n)).
 * Space complexity: O(1).
 */

public class Solution
{
    public bool IsPalindrome(int number)
    {
        // Process:
        //     1. Handle negatives immediately (return false).
        //     2. Reverse digits and compare with the original number.
        if (number < 0)
            return false;

        var mod = 10;
        var reversedNumber = 0;
        var manipulatedNumber = number;
        
        while (manipulatedNumber != 0)
        {
            reversedNumber = (reversedNumber * mod) + (manipulatedNumber % mod);
            manipulatedNumber /= mod;
        }

        return number == reversedNumber;
    }
}

File: easy/9. Palindrome Number/solution.py
# Solution for the 'Palindrome Number' problem.
# While reversing the original number is a straightforward solution, it can be less 
# performant than reversing half for large digit numbers.
#
# Time complexity: O(log(n)).
# Space complexity: O(1).

class Solution:
    mod = 10

    def isPalindrome(self, x: int) -> bool:
        reversed_half = 0
        
        # Early return for negative numbers and multiples of 10 (except zero).
        if x < 0 or (x % 10 == 0 and x != 0):
            return False

        while x > reversed_half:
            reversed_half = (reversed_half * 10) + (x % 10)
            x //= 10

        # Check if the number is equal to its reversed half (for even number of digits)
        # or if it's equal after ignoring the middle digit (odd number of digits).
        return x == reversed_half or x == reversed_half // 10
